### PR TITLE
[TIR][USMP] Added buffer info extraction pass

### PR DESCRIPTION
**Pull request by manupa-arm**
This commit adds a pass that takes the main (call graph of operators)
TIR PrimFunc and each operators also as TIR PrimFunc. The pass will
traverse through all TIR PrimFunc starting the from main. Thereafter,
it will extract information from tir.allocates. Among the information,
the liveness conflicts are reported.

* Added test for a linear model
* Added test for parallel/serial mixed for loops
* Added test for a substructure of inception-style model.
* Exposed buffer_info creation to python
* Added member functions to update pool info
* Unit tests to cover functionality of buffer_info

There is a subsequent PR depends on this : https://github.com/manupa-arm/incubator-tvm/pull/1

### CODE REVIEW

**Comment by manupa-arm** [comment]
I think this could be reviewed now. @areusch @jroesch @jcf94

**Comment by junrushao1994** [comment]
Hey @manupa-arm, thanks for the nice PR!

As a main thing I am a little bit worried about, I would love to discuss specifically on the IR-related change, because in the worse case, adding/changing/removing an IR node would result in malfunctioning of many passes. Therefore, we tend to keep the IR stable if possible, and scrutinize any changes to the IR. For example, when @masahi was adding the while node (really awesome work https://github.com/apache/tvm/pull/7425), we discussed very carefully on the implication, and actually updated many passes that are affected by that particular change.

As we are adopting the new RFC process, I would love to propose that if the RFC implies any IR modification, the modification should be written explicitly, and be discussed really carefully on everything that could happen. CC: @jroesch @vinx13 @comaniac 

On this particular case, we are seeing a set of IR changes, including this PR, the [`PrimFunc` PR](https://github.com/apache/tvm/pull/8452), the [`allocate_const` PR](https://github.com/apache/tvm/pull/8472). That will be super helpful if you guys could send a RFC discussing the rationale and potential implication on this set of changes to TIR.

🙏 Many thanks!

**Comment by manupa-arm** [comment]
Hi @junrushao1994 ,

Yes, we are working on writing up the allocate_const RFC and hoping to post it soon. For this one's addition of pinned_memory attribute to allocate, happy to write up a RFC but I feel there is an alternative w/o changing the IR if thats preferred using AttrStmt -- let me know what you think.

**Comment by tqchen** [comment]
Thanks @manupa-arm . I believe @junrushao1994 's comment was mainly on the process thing. 

It would be great if we can separate the IR node change RFC from the feature RFC(USMP), since IR changes would involve more deliberation, and explicitly marking the RFC about the IR node change would help to get the folks involved in these parts without necessarily going to deep into the feature RFC.

Back to the pinned memory itself, we might be able to fold some of that info into the PointerType storage_scope field as what @masahi did in some of his recent refactor

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (areusch) [inline src/tir/usmp/analysis/extract_buffer_info.cc:59]: per https://google.github.io/styleguide/cppguide.html#Variable_Names, add an `_` at the end

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (areusch) [inline src/tir/usmp/analysis/extract_buffer_info.cc:121]: what if pinned_memory is empty?

File: include/tvm/tir/usmp/utils.h
Comment (areusch) [inline include/tvm/tir/usmp/utils.h:47]: it seems like this could be empty; what should a planner do then? does it get a list of global memories it could pick from? is there policy defined around each memory e.g. "can_fallback" or "readable_from_targets"

**Review by areusch — COMMENTED** [review]
@manupa-arm i didn't have time to do a full review but here's some early feedback

**Comment by areusch** [comment]
i think the current status of this PR is that it's blocked on https://github.com/apache/tvm-rfcs/pull/22 . @manupa-arm do you agree?

**Comment by manupa-arm** [comment]
I think this is blocked on me doing RFC text changes in https://github.com/apache/tvm-rfcs/pull/23 and reflecting that to USMP RFC.

This one does not consider tir.allocate_const / tir.constants just yet; We are planning to add that in a seperate PR once https://github.com/apache/tvm-rfcs/pull/22 is resolved and https://github.com/apache/tvm/pull/8509 is landed.

Sorry, I ll spend the next cycles on https://github.com/apache/tvm-rfcs/pull/23 --> USMP RFC --> this one after landing this : https://github.com/apache/tvm/pull/8795 which keeps getting conflicted

**Comment by manupa-arm** [comment]
This has been updated to reflect the changes as per the discussion on the RFCs.
However, currently it depends on the PR : #9168 , therefore putting this to draft until that is merged.

**Comment by manupa-arm** [comment]
a friendly ping @mbs-octoml !

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (mbs-octoml) [reply src/tir/usmp/analysis/extract_buffer_info.cc:46]: A quick overview of the alg would be great: collect the min/max statement indexes for every buffer var occurrence, the BufferInfos for each tir.allocate, and the (unique, right?) BufferInfos for func params  from calls, then sweep the min/max intervals for the conflicts.

There's no control flow to worry about so simple min/max regions suffice.

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (mbs-octoml) [inline src/tir/usmp/analysis/extract_buffer_info.cc:185]: At most one call per function, right? Or at least if there are multiple calls then they all use the same buffers. Am I reading that right? CHECK fail if that assumption fails?

Comment (mbs-octoml) [reply src/tir/usmp/analysis/extract_buffer_info.cc:185]: this also assumes the calls are seen after the PrimFunc defns, right?

**Review by mbs-octoml — APPROVED** [review]
Sorry for the delay, finally got to this. LGTM, just some nits and a question about multiple calls to the same PrimFunc with diff buffers.

File: include/tvm/tir/usmp/utils.h
Comment (manupa-arm) [reply include/tvm/tir/usmp/utils.h]: Done

File: include/tvm/tir/usmp/utils.h
Comment (manupa-arm) [reply include/tvm/tir/usmp/utils.h]: Done

File: include/tvm/tir/usmp/utils.h
Comment (manupa-arm) [reply include/tvm/tir/usmp/utils.h:113]: Yes -- that has to be captured at the creation of the tir.allocate node though. Therefore, at the point of BufferInfo creation that is already decided.

File: python/tvm/tir/usmp/analysis/analysis.py
Comment (manupa-arm) [reply python/tvm/tir/usmp/analysis/analysis.py:25]: Oops. done :D

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc]: Done

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc]: Good point! added a description.

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc]: Done

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc:312]: I found this while integrating the USMP :) in my local. Added more code to support per-call analysis that lead sporadic liveness. Also added a test case to show this as well.

Therefore, now this pass could handle multiple calls per PrimFunc.

File: src/tir/usmp/analysis/extract_buffer_info.cc
Comment (manupa-arm) [reply src/tir/usmp/analysis/extract_buffer_info.cc]: I did move anyway :D

**Review by manupa-arm — COMMENTED** [review]
@mbs-octoml @areusch I finally got around to address the comments! PTAL when you have some time.

**Comment by manupa-arm** [comment]
@areusch a friendly ping!

File: python/tvm/tir/usmp/utils.py
Comment (areusch) [inline python/tvm/tir/usmp/utils.py:117]: note: looks like this is cruft from a previous incarnation of this PR. TODO(@manupa-arm): remove in a follow-on

**Review by areusch — APPROVED** [review]

**Pull request merged**
